activesupport/lib/active_support/file_evented_update_checker.rb

```diff
@@ -0,0 +1,71 @@
+require 'listen'
+  module ActiveSupport
+    class FileEventedUpdateChecker
+      attr_reader :listener
+      def initialize(files, dirs={}, &block)
+      	@files = Set.new
+      	@files = files.map { |f| File.expand_path(f)}
+      	@dirs = Hash.new
+      	dirs.map do |key,value|
+          @dirs[File.expand_path(key)] = Array(value) if !Array(value).empty?
+      	end
+      	@block = block
+      	@modified = false
+      	@listener = Listen.to(*base_directories, relative_paths: false)
+      	@listener.change(&method(:changed))
+      	@listener.start
+      end
+
+      def updated?
+        @modified
+      end
+
+      def execute
+        @block.call
+      ensure
+        @modified = false
+      end
+
+      # Execute the block given if updated.
+      def execute_if_updated
+        if updated?
+          execute
+          true
+        else
+          false
+        end
+      end
+
+      def watching?(file)
+        return true if @files.include?(file)
+        cfile = file
+        while !cfile.eql? "/"
```

Comment:
This loop seems suspicious to me, looks too complicated and probably non-portable. Let's see if we can simplify it.

Since you store expanded paths, wouldn't it be the case that a file is watched if it belongs to `@files` or else if one of the watched dir names is a prefix of the file name, and in that case if the extension of the file is among the associated ones for that directory? Off the top of my head.